### PR TITLE
Add schema for recent projects

### PR DESCRIPTION
**Pull request by k9845**
Addresses https://github.com/the-deep/server/issues/914

## Changes
- Added query to resemble with rest `projects-stat/recent`
Mention related users here if any.

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [ ] permission checks (tests here too)
- [ ] translations

### CODE REVIEW

File: apps/entry/factories.py [hunk 1]
```diff
@@ -3,14 +3,15 @@
...
-
+from analysis_framework.factories import AnalysisFrameworkFactory
 from .models import Entry
 
 
 class EntryFactory(DjangoModelFactory):
     entry_type = fuzzy.FuzzyChoice(Entry.TagType)
     excerpt = fuzzy.FuzzyText(length=100)
     image = factory.SubFactory(FileFactory)
+    analysis_framework = factory.SubFactory(AnalysisFrameworkFactory)
```
Comment (thenav56) [inline apps/entry/factories.py]: Let's pass this manually. Basically, entries will not be listed in graphql if the AF is different from projects' AF.

File: apps/project/schema.py [hunk 2]
```diff
@@ -34,13 +65,23 @@ class Meta:
...
+    number_of_leads = graphene.Field(graphene.Int)
+    number_of_leads_tagged = graphene.Field(graphene.Int)
+    number_of_leads_tagged_and_verified = graphene.Field(graphene.Int)
+    number_of_entries = graphene.Field(graphene.Int)
+    number_of_users = graphene.Field(graphene.Int)
+    leads_activity = graphene.List(DateCountType)
+    entries_activity = graphene.List(DateCountType)
+    activity_log = generic.GenericScalar()
+    top_sourcers = graphene.List(EntityDataType)
+    top_taggers = graphene.List(EntityDataType)
```
Comment (thenav56) [inline apps/project/schema.py]: Let's move this to stats = graphene.Field(ProjectStatType)

And define resolve_stats (here use dataloader to calculate annotate stats) and move others resolves there.

Basically, we don't want to calculate all the time, only when users ask for them.

File: utils/graphene/fields.py [hunk 3]
```diff
@@ -259,3 +259,15 @@ def resolve_url(root, info, **kwargs) -> str:
...
+class DateCountType(graphene.ObjectType):
+    date = graphene.String()
+    count = graphene.Int()
+
+
+class EntityDataType(graphene.ObjectType):
+    id = graphene.String()
+    name = graphene.String()
+    user_id = graphene.String()
+    count = graphene.Int()
```
Comment (thenav56) [inline utils/graphene/fields.py]: Maybe change name to UserEntityCountType?

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Review by bimal125 — APPROVED** [review]

File: apps/entry/factories.py [hunk 4]
```diff
@@ -3,7 +3,6 @@
 from factory.django import DjangoModelFactory
 
 from gallery.factories import FileFactory
-
```
Comment (thenav56) [inline apps/entry/factories.py]: Please add this line. Separate absolute and relative imports.

File: apps/project/schema.py [hunk 5]
```diff
@@ -52,6 +105,24 @@ def resolve_allowed_permissions(root, info) -> List[str]:
...
+
+    @staticmethod
+    def resolve_top_sourcers(root, info, **kwargs):
+        from lead.models import Lead
+        return get_top_entity_contributor(root, Lead)
+
+    @staticmethod
+    def resolve_top_taggers(root, info, **kwargs):
+        from entry.models import Entry
+        return get_top_entity_contributor(root, Entry)
```
Comment (thenav56) [inline apps/project/schema.py]: Use dataloader when there is n+1.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/project/schema.py (hunk 5 above)
Comment (k9845) [reply apps/project/schema.py]: Moving this to `ProjectDetailType` and using that in `recent_projects`

File: apps/project/dataloaders.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            key: Cast(KeyTextTransform(key, 'stats_cache'), models.IntegerField())
+            for key in [
+                ('number_of_leads'),
+                ('number_of_leads_tagged'),
+                ('number_of_leads_tagged_and_verified'),
+                ('number_of_entries'),
+                ('number_of_users'),
+                # NOTE: Used for sorting in discover projects
+                ('leads_activity'),
+                ('entries_activity'),
```
Comment (thenav56) [inline apps/project/dataloaders.py]: These are not required here.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/project/tests/test_schemas.py [hunk 7]
```diff
@@ -99,6 +194,49 @@ def test_projects_query(self):
...
+                }
+              }
+          '''
+
+        user = UserFactory.create()
+        analysis_framework = AnalysisFrameworkFactory.create()
+        public_project1 = ProjectFactory.create(title="P1")
+        public_project2 = ProjectFactory.create(title="P2")
+        public_project3 = ProjectFactory.create(title="P3")
+        public_project4 = ProjectFactory.create(title="P4")
```
Comment (thenav56) [inline apps/project/tests/test_schemas.py]: Use `create_batch`.

File: apps/project/tests/test_schemas.py [hunk 8]
```diff
@@ -23,12 +31,92 @@ def test_project_query(self):
...
+                  }
+                }
               }
             }
         '''
 
         user = UserFactory.create()
         public_project = ProjectFactory.create()
+        public_project2 = ProjectFactory.create()
+        public_project3 = ProjectFactory.create()
```
Comment (thenav56) [inline apps/project/tests/test_schemas.py]: Use `create_batch`.

**Review by thenav56 — CHANGES_REQUESTED** [review]

**Pull request merged**